test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,456 @@
+
+                try:
+                    cpp_test_name = '{}_test_{}'.format(module_variant_name, method_name)
+                    cpp_test_fn = getattr(cpp_module, cpp_test_name)
+                    if not test_params.has_parity:
+                        with self.assertRaisesRegex(RuntimeError, "Parity test failed"):
+                            cpp_test_fn(*cpp_args)
+                    else:
+                        cpp_test_fn(*cpp_args)
+                finally:
+                    # Ideally we would like to not have to manually delete the file, but NamedTemporaryFile
+                    # opens the file, and it cannot be opened multiple times in Windows. To support Windows,
+                    # we close the file after creation and try to remove it manually.
+                    for module_file_name in module_file_names:
+                        try:
+                            os.remove(module_file_name)
+                        except OSError as e:
+                            warnings.warn("Unable to remove {}, got error: {}".format(module_file_name, str(e)))
+
+        test_methods(test_params)
+
+
+def _process_test_params(test_params_dict, module_metadata, device):
+    module_name = test_params_dict.get('module_name')
+    desc = test_params_dict.get('desc', None)
+    module_variant_name = module_name + (('_' + desc) if desc else '') + (('_' + device) if device != 'cpu' else '')
+    input_size = test_params_dict.get('input_size', None)
+    input_fn = test_params_dict.get('input_fn', None)
+    if input_size:
+        example_inputs = [torch.randn(input_size)]
+    elif input_fn:
+        example_inputs = list(input_fn())
+    else:
+        raise RuntimeError("Missing `input_size` or `input_fn` for {}".format(module_variant_name))
+    if device != 'cuda' or TEST_CUDA:
+        example_inputs = [x.to(device) for x in example_inputs]
+    return TorchNNTestParams(
+        module_name=module_name,
+        module_variant_name=module_variant_name,
+        python_constructor_args=test_params_dict.get('constructor_args'),
+        cpp_constructor_args=test_params_dict.get('cpp_constructor_args'),
+        example_inputs=example_inputs,
+        has_parity=test_params_dict.get('has_parity', True),
+        python_module_class=getattr(torch.nn, module_name),
+        cpp_sources=module_metadata.get('cpp_sources', ''),
+        device=device,
+    )
+
+
+def add_test(test_name, test_fn):
+    if hasattr(TestCppApiParity, test_name):
+        raise RuntimeError("Found two tests with the same name: " + test_name)
+    setattr(TestCppApiParity, test_name, test_fn)
+
+torch_nn_modules.module_metadata_map['SampleModule'] = sample_module.module_metadata
+
+torch_nn_test_params_map = {}
+
+torch_nn_module_names = set()
+
+for test_params_dict in sample_module.module_tests + common_nn.module_tests:
+    module_name = test_params_dict.get('module_name')
+    if module_name in torch_nn_has_parity:
```

Comment:
nit: if you wanted to avoid the indentation, one could do something like

```
if module_name not in torch_nn_has_parity:
   continue
# rest of the code here
```
This helps make the code nicer when there are a lot of indentation levels, but completely up to you